### DISCUSSION TITLE
Refactoring of the triangular MPO solvers

### DISCUSSION THREAD

**Discussion by ianmccul** [mptoolkit/mptoolkit#8] [Documentation]
commit 316b3da5399946d1f7311a9e76c308d6c69ee6a8 refactors the triangular MPO solver code into a few different files.

There are now 3 'flavours' of triangular MPO solver:

- `SolveFirstOrderMPO_[Left|Right]`: this is what used to be called SolveSimpleMPO.  Solves an MPO that is first order (or first degree), meaning it has no internal identity or unitary elements on the diagonal.  These algorithms are just for demonstration purposes, as no tools actually use them.
- `SolveHamiltonianMPO_[Left|Right]`: This solves an arbitrary degree MPO, but it must have no momentum, and the degree > 1 components should vanish.  The only allowed operators on internal diagonals are the identity, or an operator that has largest eigenvalue absolute value strictly less than 1.  If there are components of degree > 1 that are larger than epsilon then this function aborts.  If there are component of degree 1 isn't proportional to the identity then this function gives a warning.
- `SolveMPO_Left`: The full solver, including momentum.  This only exists in the 'left' variant.

In all cases, there is no particular requirement that the MPS is in canonical form, as long as you supply the appropriate identity and rho operators.  For SolveHamiltonianMPO  and SolveFirstOrderMPO this requires setting the E.front() or F.back() element to the correct identity operator.

Still to be done: there is a lot of code duplication between SolveMPO_Left and SolveMPO_Left_Cross.  We can probably drop SolveMPO_Left and implement it in terms of SolveMPO_Left_Cross.

**Comment by jesse-osborne** [comment]
I've refactored the MPO solvers in the ibc-tdvp branch for solving MPOs for excitation ansatz wavefunctions.

The functions are
- `SolveFirstOrderMPO_EA_[Left|Right]`, for first order Hamiltonians.
- `SolveStringMPO_EA_[Left|Right]`, for string operators.

In the future, `SolveFirstOrderMPO_EA` could probably be generalised to handle higher-order MPOs.
